Documentation/proposals/202310-automated-sharding.md

```diff
@@ -0,0 +1,164 @@
+    # if not specified, the operator picks up the retention period if it's defined,
+    # otherwise (e.g. size-based retention) the statefulset lives forever.
+    retain:
+      retentionPeriod: 3d
+```
+
+Inspired by the [StatefulSet API](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#persistentvolumeclaim-retention), when setting `.sharRetentionPolicy.whenScaled` is set to `Delete` the operator would simply delete the underlying StatefulSet on the spot, not caring about the data stored in that particular shard. Although not the safest approach, it is exactly the current behavior, and changing it might surprise a lot of users.
+
+If `.shardRetentionPolicy.whenScaled` is set to `Retain` Prometheus-Operator will keep the shard running for a customizable period of time. By default, the Operator will keep the Shard for the whole retention period, or forever if no retention period is set. Alternatively, the user can choose a customized retention period.
+
+If retention needs to be done, the deletion logic is controlled by an annotation present on the StatefulSet:
+
+```yaml
+operator.prometheus.io/deletion-timestamp: X
+```
+
+On scale downs, the configuration of all shards would be re-arranged to make sure the Prometheus that was scaled down doesn't scrape any metrics anymore, but it would still be available for queries.
+
+When the current time finally surpasses the one in the annotation, the Operator will finally trigger the deletion of the statefulset associated with the shard.
+
+### Scaling up after scaling down
+
+If a scale-up event happens while Prometheus Servers that were marked for deletion are still running, the already running Prometheus will be unmarked and the configuration of the shard pool will be reloaded as usual.
+
+We intentionally don't want to spin up new instances while others that are marked for deletion are still running as shard names and containers' environment variables that configure each shard's target group will be too complex to coordinate.
+
+## Graceful scale-down of Prometheus Agents
+
+Prometheus Agents are different than servers since queries are not available in this mode. Their only responsibility is scraping metrics and pushing them via remote-write to a long-term storage backend, making the scale-down experience much easier to handle.
+
+When receiving the SIGTERM signal, the Prometheus Agent will gracefully handle the signal by finishing all remote-write queues before ending the process. [When terminating a Pod, Kubernetes sends the SIGTERM signal and, by default, waits 30 seconds for all containers in the pod to finish their processes](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination). Thirty seconds is usually plenty of time to finish the remote-write queues, but the PrometheusAgent CRD will be extended to allow changing the Pod's `terminationGracePeriodSeconds` field so bigger or unstable instances can be shut down without data loss.
```

Comment:
The operator sets a default termination grace period of 600s which should be more than enough. If really needed, users can use strategic merge patch to adjust the value.

Reply:
ah good reminder! I'll rephrase this part